src/view/com/composer/SelectMediaButton.tsx

```diff
@@ -0,0 +1,495 @@
+        continue
+      }
+    }
+
+    if (type === 'gif') {
+      if (isNative) {
+        errors.add(SelectedAssetError.NoGifsOnNative)
+        continue
+      }
+    }
+
+    /*
+     * All validations passed, we have an asset!
+     */
+    supportedAssets.push({
+      mimeType,
+      ...asset,
+    })
+  }
+
+  if (selectableAssetType === 'image') {
+    if (supportedAssets.length > selectionCountRemaining) {
+      errors.add(SelectedAssetError.MaxImages)
+      supportedAssets = supportedAssets.slice(0, selectionCountRemaining)
+    }
+  } else if (selectableAssetType === 'video') {
+    if (supportedAssets.length > 1) {
+      errors.add(SelectedAssetError.MaxVideos)
+      supportedAssets = supportedAssets.slice(0, 1)
+    }
+
+    const selectedVideo = supportedAssets[0]
+
+    if (typeof selectedVideo.duration !== 'number') {
+      try {
+        const metadata = await getAdditionalVideoMetadata(selectedVideo)
+        selectedVideo.duration = metadata.duration
+        selectedVideo.width = metadata.width
+        selectedVideo.height = metadata.height
+      } catch (e: any) {
+        logger.error(`processSelectedAssets: failed to get video metadata`, {
+          safeMessage: e.message,
+        })
+        errors.add(SelectedAssetError.Unsupported)
+        supportedAssets = []
+      }
+    } else {
+      /*
+       * The `duration` is in seconds on web, but in milliseconds on
+       * native. We normalize to milliseconds.
+       */
+      if (isWeb) {
+        selectedVideo.duration = selectedVideo.duration * 1000
+      }
+    }
+
+    if (
+      selectedVideo.duration &&
+      selectedVideo.duration > VIDEO_MAX_DURATION_MS
+    ) {
+      errors.add(SelectedAssetError.VideoTooLong)
+      supportedAssets = []
+    }
+  } else if (selectableAssetType === 'gif') {
+    if (supportedAssets.length > 1) {
+      errors.add(SelectedAssetError.MaxGIFs)
+      supportedAssets = supportedAssets.slice(0, 1)
+    }
+  }
+
+  return {
+    type: selectableAssetType!, // set above
+    assets: supportedAssets,
+    errors,
+  }
+}
+
+export function SelectMediaButton({
+  disabled,
+  allowedAssetTypes,
+  selectedAssetsCount,
+  onSelectAssets,
+}: SelectMediaButtonProps) {
+  const {_} = useLingui()
+  const {requestPhotoAccessIfNeeded} = usePhotoLibraryPermission()
+  const {requestVideoAccessIfNeeded} = useVideoLibraryPermission()
+  const sheetWrapper = useSheetWrapper()
+  const t = useTheme()
+
+  const selectionCountRemaining = MAX_IMAGES - selectedAssetsCount
+
+  const processSelectedAssets = useCallback(
+    async (rawAssets: ImagePickerAsset[]) => {
+      const {
+        type,
+        assets,
+        errors: errorCodes,
+      } = await processImagePickerAssets(rawAssets, {
+        selectionCountRemaining,
+        allowedAssetTypes,
+      })
+
+      /*
+       * Convert error codes to user-friendly messages.
+       */
+      const errors = Array.from(errorCodes).map(error => {
```

Comment:
would these errors show Toast messages to the user to say why their files failed to upload? 👀